joystick.c

```diff
@@ -0,0 +1,178 @@
+/**********************************************/
+// File     :     joystick.c
+// Author   :     John Bliss
+// Date     :     September 27th 2019
+/**********************************************/
+
+#include "joystick.h"
+
+
+enum joy_status joy1_mode = NONE;
+enum joy_status joy2_mode = NONE;
+
+
+static SDL_GameController *joystick1 = NULL;
+static SDL_GameController *joystick2 = NULL;
+static bool old_clock = false;
+static bool writing = false;
+static uint16_t joystick1_state = 0;
+static uint16_t joystick2_state = 0;
+static uint8_t clock_count = 0;
+
+
+bool joystick_latch, joystick_clock;
+bool joystick1_data, joystick2_data;
+
+bool joystick_init()
+{
+	int joystick1_number = 0;
+	//Try to get first controller, if it is not set to 1
+	if (joy1_mode != NONE) {
+		for (int i = 0; i < SDL_NumJoysticks(); i++) {
+			if (SDL_IsGameController(i)) {
+				joystick1 = SDL_GameControllerOpen(i);
+				if (joystick1) {
+					joystick1_number = i;
+					break;
+				} else {
+					fprintf(stderr, "Could not open gamecontroller %i: %s\n", i, SDL_GetError());
+				}
+			}
+		}
+	}
+	if (joy2_mode != NONE) {
+		for (int i=0; i < SDL_NumJoysticks(); i++) {
+			if (SDL_IsGameController(i) && joystick1_number != i) {
+				joystick2 = SDL_GameControllerOpen(i);
+				if (joystick2) {
+					break;
+				} else {
+					fprintf(stderr, "Could not open gamecontroller %i: %s\n", i, SDL_GetError());
+				}
+			}
+		}
+	}
+	writing = false;
+	return true;
+}
+
+void joystick_step()
+{
+	if (joy1_mode == NONE) {
+		return;
+	}
+	if (!writing) { //if we are not already writing, check latch to
+		//see if we need to start
+		handle_latch(joystick_latch, joystick_clock);
+	}
+
+	//if we have started writing controller data and the latch has dropped,
+	// we need to start the next bit
+	if (writing & !joystick_latch) {
```

Comment:
The `writing` variable is tested immediately above here; it doesn't need to be tested again.  Therefore, the line would be better as
```c
else if (!joystick_latch) {
```

Reply:
I have updated this.